closed/src/java.base/share/classes/com/sun/crypto/provider/NativeGaloisCounterMode.java

```diff
@@ -469,39 +471,50 @@ int decryptFinal(byte[] in, int inOfs, int len,
                      byte[] out, int outOfs)
         throws IllegalBlockSizeException, AEADBadTagException,
         ShortBufferException {
+        // avoid any race conditions in OpenSSL when the key/iv are used concurrently
+        byte[] keyCopy = key.clone();
+        byte[] ivCopy = iv.clone();
+
+        // need to synchronize this block as if there are concurrent calls to decryptFinal
+        // the shared variables aadBuffer and ibuffer may corrupt input and cause a segfault
+        // in openssl code
         if (len < 0) {
             throw new ProviderException("Input length is negative");
         }
-        if (len < tagLenBytes - ibuffer.size()) {
+
+        int bufSize = 0;
+        synchronized (this) {
+            bufSize = ibuffer.size();
+            if (len > 0) {
+                ibuffer.write(in, inOfs, len);
```

Comment:
`ibuffer` should not be modified if an exception is thrown below (e.g. line 476).

Reply:
Would it matter if the `ibuffer` was modified? It will be reset afterwards on [L494](https://github.com/ibmruntimes/openj9-openjdk-jdk16/blob/cc76111e5443112bba3dd3b09aae3456299fd582/closed/src/java.base/share/classes/com/sun/crypto/provider/NativeGaloisCounterMode.java#L494).

Reply:
This has not yet been addressed.

Reply:
Okay, I moved it till after the checks.